air/src/execution_step/lambda_applier/applier.rs

```diff
@@ -104,13 +98,44 @@ fn select_by_path_from_stream<'value>(
     Ok(select_result)
 }
 
+fn select_by_path_from_canon_map_stream<'value>(
+    stream: impl ExactSizeIterator<Item = (&'value JValue, RcSecurityTetraplet)> + 'value,
+    lambda: &NonEmpty<ValueAccessor<'_>>,
+    exec_ctx: &ExecutionCtx<'_>,
+) -> ExecutionResult<MapLensResult<'value>> {
+    let stream_size = stream.len();
+    let (idx, body) = extract_idx(lambda, exec_ctx)?;
+
+    let (value, tetraplet) = lambda_to_execution_error!(stream
+        .peekable()
+        .nth(idx)
+        .ok_or(LambdaError::CanonStreamNotHaveEnoughValues { stream_size, idx }))?;
+
+    let select_result = if body.is_empty() {
+        // csm.$.key.[0] case
+        let result = Cow::Borrowed(value);
+        MapLensResult::from_cow(result, tetraplet)
+    } else {
+        // csm.$.key.[0].attribute case
+        let result = select_by_path_from_scalar(value, body.iter(), exec_ctx)?;
+
+        let joined = body.iter().map(ToString::to_string).collect::<Vec<_>>().join(".");
+        let json_path_suffix = format!(".{}", joined);
```

Comment:
It's an abstraction leakage, no one should know about how lens is represented internally (e.g. it's united with a dot). One should use `add_lambda` instead.

Reply:
We decided to do it in the next PR.